Bluetooth: controller: Add support for initiator on secondary adv channels

andrzej-kaczmarek opened the pull request:
This adds support to initiate connection on secondary adv channels. Should pass all LL/CON/INI/* test cases.

Fixes #3514

andrzej-kaczmarek committed:
Bluetooth: controller: Refactor CONNECT_IND prepare to separate func

We can use the same code to create both CONNECT_IND and AUX_CONNECT_REQ
since they are basically that same PDUs.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add public helpers for scan checks

We will need the same checks for AdvA and TargetA/InitA in lll_scan_aux
so let's make them public and use address explicitly instead of passed
via pdu.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Use proper transmitWindowDelay for central

On secondary advertising channel transmitWindowDelay depends on PHY
used to create connection so we need to adjust it.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix aux scanning for connectable ADV_EXT_IND

When in initiating state we need to handle ADV_EXT_IND as in regular
scan sice we always want to scan AUX_ADV_IND in order to be able to
connect.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add support for initiator on secondary adv ch

This allows to initiate a connection on secondary advertising channels,
i.e. when advertises uses advertising extensions.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix 1st connection event offset

We need to take RX chain delay and PHY used to send CONNECT_IND into
account when calculating 1st connection event offset.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
The only problem I noticed on qualification test cases is that 1st connection event is sometimes sent too early - usually several usecs. Initiator on legacy adv has the same problem so seems like some calculations are a bit off, but I was not able to figure it out yet.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -819,73 +907,26 @@ static inline int isr_rx_pdu(struct lll_scan *lll, struct pdu_adv *pdu_adv_rx,
-				CONN_INT_UNIT_US);
-			pdu_tx->connect_ind.win_size++;
+#endif
+			init_tx_addr = pdu_adv_rx->rx_addr;
```
```suggestion
			init_tx_addr = lll->init_addr_type;
```

cvinayak reviewed (changes requested):
Good work. Minor regression otherwise looks good to me.

cvinayak committed:
Bluetooth: controller: Ensure only initiating on selected PHY

Reset the scanning context for the PHY not selected in the
Extended Create Connection command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fixes and use of defines related to initiator

Fixed incorrect LLL context used to generate done event for
auxiliary channel scanning. Other minor comments and fixes.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Reserve an additional node rx for Ext. Initiator

Reserve an additional node rx buffer when Extended Initiator
is supported as the received ADV_EXT_IND PDU is being
buffered in the auxiliary channel scanning instance and is
only released/flushed in the done event of the initiating
auxiliary channel radio event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Implement disabling the other PHY initiator

When connection is initiated in one of either 1M or Coded
PHY initiating scan instance then the other scanning
instance's scheduling and memory allocation needs to be
cleaned up.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Ensure both 1M and Coded PHY can initiate

Ensure that the advertiser address type and address is setup
in both 1M and Coded PHY scanning instance when extended
create connection is enabled on both the PHYs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Rename use of params to param

Rename the use of params word with param.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Do not report Aux PDUs while initiating

Do not report auxiliary PDUs as advertising reports when in
initiator state on auxiliary channels.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
I have discovered during testing that there has always been a bug during connection establishment when it comes to stopping the initiating scan events, continuous scanning could use prepare pipeline and the pipeline is not flushed under this scenario. Now I remember, the previously used `is_stop` flag in ULL hdr was getting used such that the scan events in pipeline got aborted, but this flag no longer exists (not that I want it back, it was incorrect to use it in first place).

The correct fix will be:
1. Remove the `is_enabled = 0U;`  reset when handling `NODE_RX_TYPE_CONNECTION` 
2. Remove the reset of `lll_scan->conn = NULL;` being done in `ull_master_setup`. This will allow scan events to abort early when connection has already been aborted
3. Hold on to `NODE_RX_TYPE_CONNECTION` from being available to LL context until the scan event is done and `disabled_cb` is called.

Will think more on this.

andrzej-kaczmarek committed:
Bluetooth: host: Add adv parameters def for connectable ext adv

This is the same as for legacy, but with ext adv flag.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
tests: Bluetooth: bsim: Fix conn ref

'disconnected' callback always unrefs conn object, but this is only
valid if connection was created in scanx since app owns reference to
that conn object. In case of advx, app does not own any reference so
should not unref.

Basically we do not really need to hold any reference to conn object
in either scanx or advx so just drop reference immediately after conn
is created in scanx, no need to worry about it later.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
tests: Bluetooth: bsim: Handle consecutive connection in tests

If we want to test conn on legacy and ext adv, we need scanx to
handle connection more than once. Change local flag to global and
let it indicate whether we want scanx to initaite connection when
device is scanned.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
tests: Bluetooth: bsim: Add test for initaiting conn on ext adv

advx will do connectable advertising using ext adv, scanx will try
to connect.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak reviewed (approved):
Have tested on target to satisfactory results. LGTM,

cvinayak commented:
@andrzej-kaczmarek please rebase and push to trigger CI, I think the build failure is unrelated.

carlescufi reviewed (approved):
LGTM
